src/alloc.c

```diff
@@ -1011,14 +1015,32 @@ JL_DLLEXPORT jl_datatype_t *jl_new_bitstype(jl_value_t *name, jl_datatype_t *sup
 
 // type constructor -----------------------------------------------------------
 
-jl_typector_t *jl_new_type_ctor(jl_svec_t *params, jl_value_t *body)
+JL_DLLEXPORT jl_value_t *jl_new_type_constructor(jl_svec_t *p, jl_value_t *body)
 {
-    jl_typector_t *tc = (jl_typector_t*)newobj((jl_value_t*)jl_typector_type,NWORDS(sizeof(jl_typector_t)));
+    size_t i, np = jl_svec_len(p);
+    jl_value_t **env;
+    JL_GC_PUSHARGS(env, np * 2 + 2);
+    jl_svec_t *params = jl_alloc_svec(np);
+    env[np * 2] = (jl_value_t*)params;
+    for (i = 0; i < np; i++) {
+        // recreate body with the TypeVar unbound
+        jl_tvar_t *tv = (jl_tvar_t*)jl_svecref(p, i);
+        assert(jl_is_typevar(tv));
+        env[2 * i] = (jl_value_t*)tv;
+        tv = jl_new_typevar(tv->name, tv->lb, tv->ub);
```

Comment:
Did you mean `0` instead of `tv->ub`? Otherwise I'm not sure what this is doing. Or we could have the front end pass in unbound typevars in the first place I think.